### PR TITLE
Add namespace feature to model / and simplify model.py

### PR DESCRIPTION
**Pull request by aseyboldt**
Like #4123, but removes a lot of very complicated code in model.py, and uses namespaces to get (most?) of the original functionality.
I'm not sure if this is a good idea as it is, just for discussion.

### CODE REVIEW

File: pymc3/model.py [hunk 1]
```diff
@@ -882,75 +661,80 @@ def __init__(self, mean=0, sigma=1, name='', model=None):
...
-            theano_config = {"compute_test_value": "raise"}
-        instance._theano_config = theano_config
-        return instance
-
-    def __init__(self, name="", model=None, theano_config=None, coords=None):
+        parent = kwargs.get('model', None)
+        if parent is None:
+            parent = cls.get_context(error_if_none=False)
+        if parent is None:
+            return super().__new__(cls)
```
Comment (AlexAndorra) [inline pymc3/model.py:704]: The `if` condition is duplicated here -- or maybe the `else` is missing ?

**Comment by michaelosthege** [comment]
I like this. It simplifies the code quite a lot.
We'll need to...
+ [ ] resolve conflicts
+ [ ] fix the tests
+ [ ] make a list of the breaking changes

Even if we don't want to do the namespace stuff (which we already do), we should take these simplifications!

**Comment by twiecki** [comment]
@aseyboldt any progress on this? Would be a great feature.

**Comment by canyon289** [comment]
Figure we wont get back to this until v4 so marking as such

**Comment by michaelosthege** [comment]
Like #4123 I'm closing this because in the meantime we made major changes to the internals.

I'm not opposed to the namespace idea, but we should maybe re-evaluate it when we all have more routine with `v4`.

**Pull request closed without merging**